### PR TITLE
DMD-1939: agent provisioning support (create_project + bootstrap sessions)

### PR DESCRIPTION
**Pull request by vojtabiberle**
## Description

**Linear**: [DMD-1939](https://linear.app/keboola/issue/DMD-1939/agent-provisioning-support-in-keboolamcp-server)

**RFC**: [`feature_spec/agent_provisioning/RFC.md`](feature_spec/agent_provisioning/RFC.md)

### Change Type

- [ ] Major (breaking changes, significant new features)
- [x] Minor (new features, enhancements, backward compatible)
- [ ] Patch (bug fixes, small improvements, no new features)

### Summary

Makes the Connection agent-provisioning endpoint from [DMD-1806](https://linear.app/keboola/issue/DMD-1806/variant-c-agent-created-project-with-immediate-user-confirmation) (`POST /manage/programmatic-projects`, keboola/connection#8081) reachable from an MCP client, so a first-time user gets a working Keboola project without leaving the conversation. Until now the server needed an existing project **and** a token, and refused to start a session without one.

**`create_project` tool.** Provisions a project under the stack's agent maintainer and signs the session in to it. Returns the project id/name/backend, the human's `confirm_url` and a backend-init flag. The access and refresh tokens are **never** part of the result — they would land in the model context and the transcript — they go to the local credential store only. Optional `name` and `backend` (`snowflake`/`bigquery`, null keeps the maintainer default); the request carries a `clientId` derived from the MCP client's name, so `auditLog.agentProvisioning.projectProvisioned` gets real attribution.

**Storing the session is what makes the next call work.** The endpoint issues the same AT/RT programmatic session the PKCE login does, so it is written as an ordinary `TokenSet` and the existing code does the rest: `_maybe_use_stored_session` picks it up on the next request, `get_access_token` refreshes it at the already-existing `POST /v1/auth/token/refresh` before the 1 h access token expires, and `project_ids=[id]` makes `_read_persisted_login_scope` return a *confirmed* single-project scope, so data tools work immediately instead of being held at the ask-first gate.

**Bootstrap sessions.** `create_session_state` no longer raises on a missing token; it returns a state with no `KeboolaClient`, and the server starts, serves `tools/list` and runs `create_project` with only a stack URL. Tools that need Keboola access get an actionable message from `KeboolaClient.from_state` (one guard in the shared accessor, not ~40 call sites) instead of a `KeyError`. `_local_login_fallback` no longer fails stdio startup when nothing is stored.

**Revoked-session recovery.** Confirming the claim revokes the agent session, so its access token 401s while still unexpired and the expiry-driven refresh never runs. A 401 (walked through the `tool_errors` cause chain) now re-checks the credential with `introspect_token` and drops it only when that fails too — a scope-related 401 keeps a valid credential, which is the common case today.

**Error mapping.** A stack without the `agent-provisioning` feature (404) reports "not available on this stack", no stack trace; a 429 keeps its `Retry-After`.

**Explicitly out of scope** (see the RFC): `syncBackendInit` (the endpoint defaults to async and a chat session should not block for minutes on backend init), provisioning from an already-credentialed session (refused — it would overwrite the login that session is using), and provisioning on the deployed server (OAuth, no local credential store — refused with a message).

## Testing

- [ ] Tested with Cursor AI desktop (`Streamable-HTTP` transports)

### Optional testing
- [ ] Tested with Cursor AI desktop (all transports)
- [ ] Tested with claude.ai web and `canary-orion` MCP (`Streamable-HTTP`)
- [ ] Tested with In Platform Agent on `canary-orion`
- [ ] Tested with RO chat on `canary-orion`

Unit tests cover the endpoint mapping (including `accessTokenExpiresIn`, 404 and 429), the tool (no token material in the serialized result, the stored session and its scope, `backend`/`clientId` forwarding, and every refusal), bootstrap session state and `tools/list` without a client, and the revoked-vs-scope 401 distinction. Full `tox` passes (2199 pytest, ruff, check-tools-docs).

**Not yet run against a real stack.** The end-to-end flow needs a stack with `agent-provisioning` enabled (canary-orion): start the server with only `KBC_STORAGE_API_URL`, call `create_project`, confirm the URL in a browser, and check that a data tool works before confirmation and reports an ended session after it.

## Checklist

- [x] Self-review completed
- [x] Unit tests added/updated (if applicable)
- [ ] Integration tests added/updated (if applicable)
- [x] Project version bumped according to the change type
- [x] Documentation updated (if applicable)

### CODE REVIEW

File: src/keboola_mcp_server/mcp.py [hunk 1]
```diff
@@ -389,11 +419,43 @@ async def on_request(
...
+            return
+        storage_api_url = cast(str, config.storage_api_url)
+        if load_tokens(storage_api_url) is None:
+            return
+        try:
+            await introspect_token(storage_api_url, subject_token=strip_bearer(cast(str, config.storage_token)))
+            return  # the credential itself is fine; the 401 was about something else
+        except Exception:
+            LOG.info(f'The stored session for {storage_api_url} is no longer valid; dropping it.')
+        forget_tokens(storage_api_url)
```
[the remark this answers is not in the document]
Comment (vojtabiberle) [reply src/keboola_mcp_server/mcp.py]: Good catch on all three — fixed in 26757c75.

The deletion now requires all of these, and `_handle_unauthorized` documents them:

1. **The request used the stored token.** `config.storage_token` can be a per-request header token; its 401 says nothing about the stored session. Compared against `load_tokens(...).access_token` before anything else happens, so a header token never reaches the introspection call at all.
2. **Introspection fails with 401/403.** A timeout, connection error or 5xx is logged as "could not verify; keeping it" and the credential stays — only an authentication failure counts as evidence.
3. **Compare-and-delete.** The removal moved into `auth_login.forget_rejected_access_token()`, which re-reads the entry under the same per-(stack, profile) `asyncio.Lock` + `flock` that `get_access_token()` refreshes under, and deletes only while the stored access token is still the rejected one. A concurrent refresh or a fresh `login` in between is left alone. Same shape as the existing dead-refresh-token handling in `get_access_token()`.

Covered by `TestUnauthorizedDropsARevokedSession` (dropped for 401/403; kept for a transport error, a 503, a per-request token, and a scope-related 401 that introspection survives) and `test_forget_rejected_access_token_compares_before_deleting` (same token, same token with a `Bearer ` scheme, rotated meanwhile, never-stored token).

File: src/keboola_mcp_server/tools/project.py [hunk 2]
```diff
@@ -664,3 +682,121 @@ async def set_project_scope(
...
+    backend_init_pending: bool = Field(
+        description=(
+            'True when the storage backend is still being initialised in the background, so the '
+            'first data operations may fail for a short while.'
+        )
+    )
+    llm_instruction: str = Field(description='What to tell the user next.')
+
+
+@tool_errors()
```
[the remark this answers is not in the document]
Comment (vojtabiberle) [reply src/keboola_mcp_server/tools/project.py]: Right about the mechanism; fixed in 26757c75 by taking the "explicitly handle it" option.

`_trigger_event` now returns early when the session state has no `KeboolaClient`, with a debug line. The event is not recoverable in a bootstrap session rather than merely awkward: a Storage event is written *into a project* with a Storage token, and at that point there is neither — `create_project` is the only tool that runs in that state, and before it returns there is nothing to write to. Building a client from the freshly stored session afterwards would mean extra Connection round-trips inside a `finally` block, for one event.

What the early return buys is that the miss is now silent and intentional instead of a `Failed to trigger tool event` warning on every first call, which read like a malfunction.

Attribution is not lost: the provisioning is recorded on the Connection side as `auditLog.agentProvisioning.projectProvisioned`, carrying the `clientId` the tool sends (the MCP client name, sanitized) — which is exactly what DMD-1939 asked that field for. Every tool call after provisioning has a client and emits events as usual.

Covered by `test_bootstrap_session_skips_the_event_without_warning` (the tool still succeeds and nothing at WARNING or above is logged); it fails if the guard is removed.

**Review by Matovidlo — COMMENTED** [review]
Did you test it locally that after provisioning of the project all tools are loaded and MCP works ? Simply it has to work for both local MCP server and remote MCP server. The whole session has to be reloaded and user has to be logged in and tools should be refreshed.

**Comment by vojtabiberle** [comment]
@Matovidlo Straight answer first: **no, I have not run it against a live stack yet.** I have no stack with `agent-provisioning` enabled reachable from this machine, and the PR description says so. So please read everything below as "this is what the code does and what is now proven in tests", not as "I watched it work in Cursor".

What I did add instead of hand-waving — two tests that exercise the real middleware chain, not mocks at the tool boundary (fbbf1893):

* `tests/test_server.py::TestBootstrapServerEndToEnd` — a server built from a stack URL alone, driven by a **real MCP client**: `tools/list` works with no credential, `create_project` runs through the whole middleware chain, and the advertised tool list is asserted **identical before and after**.
* `tests/test_mcp.py::TestProvisionedSessionIsPickedUpByTheNextRequest` — the **real `on_request`** over a real credential file: the very next tool call is already signed in with the provisioned token and pinned to the new project, with a confirmed scope.

## Why no session reload or tool refresh is needed

**Session.** `SessionStateMiddleware.on_request` rebuilds the session state on *every* request, and `_maybe_use_stored_session` reads the credential store each time. `create_project` writes the provisioned AT/RT there, so the next request builds a `KeboolaClient` from it — no restart, no reconnect, no re-initialize. That is the second test above.

**Logged in.** The tokens are stored with `project_ids=[<new project>]`, which makes `_read_persisted_login_scope` return a **confirmed** scope. So the ask-first gate in `MultiProjectMiddleware` lets data tools straight through — the user does not have to call `set_project_scope`, and there is no "log in again" step between `create_project` and the first query.

**Tools.** They never need refreshing, because the advertised list does not change. In bootstrap mode `ToolsFilteringMiddleware.on_list_tools` returns the superset (no client to `verify_token` with), and after provisioning the session holds a `kbc_at_` programmatic token, for which the same method *already* returned the superset before this PR. The per-call guards still enforce every feature/role/branch rule. This is deliberately the same choice as the note in `MultiProjectMiddleware.on_list_tools`: Claude Code and Cursor do not re-fetch the list after `notifications/tools/list_changed` mid-session, so anything that relies on a refresh is broken by construction. Asserted as `after == before`.

## Remote MCP server

Here I disagree with the requirement as stated, and I'd rather settle it now than ship something misleading: **the remote server refuses `create_project` on purpose** (it answers "only supported by a locally run MCP server").

1. A remote session cannot be credential-less in the first place. The deployed server authenticates every session — OAuth, or a caller-supplied programmatic token (Kai). An unauthenticated caller never reaches a tool, so the bootstrap case the tool exists for cannot arise there.
2. Even if it could, the credential store is a **process-wide file**. Writing a provisioned session into it on a shared pod would hand that session to every other session on the replica. That is a leak, not an inconvenience.

Doing it properly on remote means persisting the provisioned AT/RT **per MCP session** in the Postgres session store the OAuth flow already uses, and teaching `on_request` to read it back — a real increment, not a flag flip. Happy to do it, but it is a separate PR with its own RFC section. Tell me which you want:

* **(a)** merge this as local-only (what is here now), remote as a follow-up; or
* **(b)** hold this PR until remote support lands too.

And if you want the live check before merging either way, point me at a stack with `agent-provisioning` on (canary-orion?) plus access and I will run the full manual E2E — start with only `KBC_STORAGE_API_URL`, `create_project`, confirm in the browser, verify the tools work before confirmation and that the session ends cleanly after it — and post the result here.

**Comment by Matovidlo** [comment]
I have tried it, it generated this project

https://connection.canary-orion.keboola.dev/admin/projects/525

https://connection.canary-orion.keboola.dev/agent-project/confirm?token=kbc_apc_01a0af62-8e2e-721c-9323-6ba42c01f428_0121580f438542dda49ac0b45fd760fe319932362f55

But this is returning 404. Not sure why its not working I am unable to claim it

**Comment by vojtabiberle** [comment]
> I have tried it, it generated this project
> 
> https://connection.canary-orion.keboola.dev/admin/projects/525
> 
> https://connection.canary-orion.keboola.dev/agent-project/confirm?token=kbc_apc_01a0af62-8e2e-721c-9323-6ba42c01f428_0121580f438542dda49ac0b45fd760fe319932362f55
> 
> But this is returning 404. Not sure why its not working I am unable to claim it

I'm not sure if everything is still deployed on Orion or if somebody rewrite it with branch withouth rebase on master/main.

**Comment by Matovidlo** [comment]
so once its somewhere ping me I will retest it

**Comment by vojtabiberle** [comment]
> so once its somewhere ping me I will retest it

Is tested on orion and now is on orion correct (main) branch of connection.

You can test it by yourself

**Comment by vojtabiberle** [comment]
## The 404 was canary-orion, not this branch

Orion was not running `master`. Its connection image was `canary-orion-jirka-8444-1`, a canary tag off keboola/connection#8444, cut from master at `d1f382505b` (2026-09-15 11:21 UTC). The confirmation work — keboola/connection#8162 (DMD-1806) — merged to master at `f88fd4b1a5` on 2026-09-16 13:07 UTC, i.e. after that branch was cut and never rebased onto it. So the stack simply had no `AgentProjectConfirmPageAction` and no `POST /manage/programmatic-projects/confirm`; the page route is deliberately *not* feature-gated, so the 404 meant "code absent", not "feature off".

Redeployed master to Orion (tag `canary-orion-master-14`, kbc-stacks keboola/kbc-stacks#24725). Orion is now on `d7f8dfb358`, and `/manage/programmatic-projects/confirm` is in its manage spec.

Heads-up for whoever was testing #8444 on Orion: that tag is no longer deployed there.

## Connection side, verified on Orion

- `GET /agent-project/confirm` → 200, renders the kbc-ui bundle (`<title>Confirm project | Keboola Connection</title>`).
- `GET /agent-project/confirm?token=…` → 302 to `/agent-project/confirm` + `Set-Cookie: kbcAgentProjectClaim=…; HttpOnly; Secure; SameSite=Lax; Max-Age=259200`. Token leaves the URL, as designed.
- With the cookie present, the page's `appData` gets `"agentProjectClaimPresent":true` and `"agentProjectClaimId":"<uuid>"` — the claim id only, never the token.
- The deployed kbc-ui chunk contains `agent-project/confirm`, `agentProjectClaimPresent`, `agentProjectSignedInAs`, so no white-screen on an old bundle.
- `POST /manage/programmatic-projects/confirm` unauthenticated → 401 `Programmatic session not found.` (not 404, so `agent-provisioning` is on).

## This branch's local MCP server, against Orion

Ran the stdio server from `DMD-1939-agent-provisioning-support` (`fbbf1893`) with `--api-url https://connection.canary-orion.keboola.dev`, no `KBC_*` env, credential store isolated via `HOME`.

**Bootstrap without a token** — server started, `tools/list` returned 44 tools including `create_project`. A data tool before any credentials returns the actionable message rather than a `KeyError`:

> This session has no Keboola credentials. Call the "create_project" tool to create a new Keboola project and a session for it, or run "keboola-mcp-server login --api-url &lt;stack-url&gt;" to sign in to an existing one.

**`create_project`** → project **527**, `backend: snowflake`, `backend_init_pending: true`. No token material in the result — only `confirm_url`. The credential store got `project_ids: [527]`, `read_only: false`, plus the access and refresh tokens.

**The stored session is immediately usable** (checked on project 526 from a first run, once backend init had finished):

- `get_project_info` → 526, Snowflake, `workspace_id: 15194`, `user_role: admin`
- `get_accessible_projects` → `agent+01M2SRCADGWTPDMZDFRVFMQ8Y7@agent-provisioning.invalid`, scope `[526]`, `in_scope: true` — so data tools are not held at the ask-first gate
- `query_data` running `SELECT CURRENT_DATABASE(), CURRENT_ROLE()` → `KEBOOLA_526, KEBOOLA_WORKSPACE_762112` — real SQL against Snowflake

**A second `create_project` on the same stack refuses**, as intended, rather than overwriting the session.

**The `confirm_url` works**: 302 + claim cookie, 3-day TTL.

Two throwaway projects are left on Orion from this: **526** and **527**, both named "DMD-1939 local MCP smoke test", unclaimed. I have a live claim link for 527 but am not pasting it here — it is a bearer credential that grants project admin. Ask me and I will send it, or just provision your own.

## Not covered

- The human half — actually clicking confirm and signing in. Needs a browser; the backend half around it is verified above.
- The deployed (remote) MCP on Orion. It still runs the image from commit `33a6a3a1` (2026-07-31), so `create_project` is not there at all. Beyond deploying this branch via kbc-stacks, the tool also has a hard guard: `if deployed_sa_token_path()` — i.e. `KBC_KUBERNETES_TOKEN_PATH` — raises "only supported by a locally run MCP server", with `test_create_project_is_refused_on_the_deployed_server` pinning it. If getting both servers working is the goal of this branch, that guard and where the provisioned AT/RT would live on a remote session need a decision; a deploy alone will not do it.

**Review by Matovidlo — APPROVED** [review]
LGTM, I have tested it on canary it works fine I have found 1 unrelated bug during testing 🚀

**Comment by vojtabiberle** [comment]
## `create_project` on a stack without `agent-provisioning` (fad267d9)

The endpoint is gated by `STACK_FEATURES__AGENT_PROVISIONING` and 404s where the feature is off, so the tool has to be usable knowledge on such a stack rather than a dead end. It already mapped the 404 to "not available"; that message said nothing the user could act on, which is the actual requirement. Now:

**The tool stays listed on every stack.** Whether a stack has the feature is not knowable at `tools/list` time — it would cost a Connection round-trip on every listing, and a bootstrap session has no credential to make one with. A tool that silently isn't there is a worse answer than a sentence explaining why.

**Calling it on such a stack returns this, and no stack trace:**

```
Creating a Keboola project is not available on https://connection.<stack>.keboola.com: this stack
does not have the "agent-provisioning" feature enabled.
Recovery: Do not retry this tool on this stack -- the feature is switched off there, not temporarily
unavailable. Tell the user that this Keboola stack cannot create a project for them, and that they
can instead use a project they already have (run "keboola-mcp-server login --api-url <stack-url>",
or supply a Storage API token), create one in the Keboola UI, or ask their Keboola administrator to
enable the "agent-provisioning" feature on the stack.
```

Split the way the codebase already splits these: the exception states the fact, and a `tool_errors(recovery_instructions={AgentProvisioningUnavailableError: ...})` entry carries what to do. The "do not retry" is deliberate — the 404 is a permanent property of the stack, and without it a model will happily call the tool three more times. The tool docstring says the same, so it is in `TOOLS.md` too.

The same message covers a stack too old to have the endpoint at all; the caller does the same thing either way.

**Tests:** `test_create_project_reports_a_stack_without_agent_provisioning` pins the message and the alternatives; `TestBootstrapServerEndToEnd::test_the_tool_is_listed_but_reports_a_stack_without_the_feature` drives a real MCP client against a 404ing stack and asserts the tool *is* listed, the call returns that message, and nothing half-provisioned is left in the credential store. Full `tox` green (2225 passed).

No version bump on top of the existing 1.82.0 → 1.83.0 — same PR.

**Comment by vojtabiberle** [comment]
## Correction + follow-up: this branch is now deployed on canary-orion

My previous comment said the deployed MCP on Orion still runs the image from `33a6a3a1` (2026-07-31), and listed the remote server as untested. That is no longer true — I deployed this branch there.

**How** (no manual kbc-stacks PR needed; `release.yml` does it): pushed tag `canary-orion-v1.83.0-dev.1` at `fbbf1893`.

- release run [35322748649](https://github.com/keboola/mcp-server/actions/runs/35322748649) → success
- kbc-stacks `canary-orion` branch, commit `ff0d507f`: `dev-keboola-canary-orion/mcp-server/tag.yaml` bumped from `canary-orion-33a6a3a1…` to `canary-orion-fbbf1893…`
- `https://mcp.canary-orion.keboola.dev/` now reports `appVersion: canary-orion-fbbf1893d300e25950e45a2143cfad96786886ee`, `serverVersion: 1.83.0`

**What the running pod shows.** `mcp-server-69986b45f4-vgwt8`, started 2026-09-18T08:12:07Z, image `keboola/mcp-server:canary-orion-fbbf1893…`, and:

```
KBC_KUBERNETES_TOKEN_PATH = /var/run/secrets/connection.keboola.com/serviceaccount/token
```

Executed inside that pod:

```
deployed_sa_token_path() = /var/run/secrets/connection.keboola.com/serviceaccount/token
create_project registered: True
```

So on the deployed server the tool **is** registered and will appear in `tools/list`, but any call hits the guard and returns "Creating a Keboola project is only supported by a locally run MCP server. Sign in through this server's OAuth flow instead." The env var is the guard's only input and it is set by the chart, so this follows from the deployment itself, not from reading the code.

I have not observed the actual `tools/call` — the remote server is behind OAuth and I could not complete that flow. Everything upstream of the guard is verified above.

If "both servers working" is the goal of this branch, this is the open design question, unchanged by the deploy: a remote session has nowhere to keep the provisioned AT/RT (that is what the guard encodes), so either the remote flow stores them somewhere server-side, or `create_project` stays local-only and the remote server documents that.

**Unrelated nit spotted during the deploy:** `trigger-image-tag-update` rewrites only `image.tag` in `tag.yaml`. The neighbouring `image.commit_sha` is still `420c32bb…` from July. Nothing seems to read it, but it now contradicts the tag next to it.
